File: cli/go/rectangle.go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/goccy/go-yaml"
	"github.com/spf13/cobra"
)

type RectangleData struct {
	A float64 `yaml:"a"`
	B float64 `yaml:"b"`
	C float64 `yaml:"c"`
	D float64 `yaml:"d"`
}

func computeRectangleArea(data RectangleData) float64 {
	width := abs(data.C - data.A)
	height := abs(data.D - data.B)
	return width * height
}

func abs(x float64) float64 {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	var rootCmd = &cobra.Command{
		Use:   "rectangle [yaml-file]",
		Short: "Calculate rectangle area from YAML file",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			yamlFile := args[0]

			start := time.Now()

			fileContents, err := os.ReadFile(yamlFile)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
				os.Exit(1)
			}

			var data RectangleData
			err = yaml.Unmarshal(fileContents, &data)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error parsing YAML: %v\n", err)
				os.Exit(1)
			}

			area := computeRectangleArea(data)

			elapsed := time.Since(start)

			fmt.Printf("Rectangle area: %.2f\n", area)
			fmt.Printf("Time: %.6f ms\n", float64(elapsed.Nanoseconds())/1e6)
		},
	}

	if err := rootCmd.Execute(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

File: cli/rust/rectangle.rs
use std::fs;
use std::time::Instant;

use clap::{Arg, Command};
use serde::Deserialize;

#[derive(Deserialize)]
struct RectangleData {
    a: f64,
    b: f64,
    c: f64,
    d: f64,
}

fn compute_rectangle_area(data: &RectangleData) -> f64 {
    let width = (data.c - data.a).abs();
    let height = (data.d - data.b).abs();
    width * height
}

fn main() {
    let matches = Command::new("rectangle")
        .about("Calculate rectangle area from YAML file")
        .arg(
            Arg::new("yaml-file")
                .help("Path to YAML file containing rectangle coordinates")
                .required(true)
                .index(1),
        )
        .get_matches();

    let yaml_file = matches.get_one::<String>("yaml-file").unwrap();

    let start = Instant::now();

    let file_contents = fs::read_to_string(yaml_file).expect("Error reading file");

    let data: RectangleData = serde_yaml::from_str(&file_contents).expect("Error parsing YAML");

    let area = compute_rectangle_area(&data);

    let elapsed = start.elapsed();

    println!("Rectangle area: {:.2}", area);
    println!("Time: {:.6} ms", elapsed.as_secs_f64() * 1000.0);
}
